Some tests require numpy and fail if not present

For reason's briefly discussed in https://github.com/dask/dask/issues/7170#issuecomment-780872128, openSUSE does not have numpy for Python 3.6 in the Tumbleweed distribution anymore. At the same time, we need to have a working test suite to make sure that the built package works (without numpy functionality of course.)

Most tests of distributed involving (runtime optional) numpy are already skipped automatically if numpy is not present. But there are some ERRORs and FAILs:

```
[  150s] FAILED distributed/tests/test_client.py::test_turn_off_pickle[True] - ModuleN...
[  150s] FAILED distributed/tests/test_client.py::test_turn_off_pickle[False] - Module...
[  150s] FAILED distributed/tests/test_client.py::test_de_serialization - ModuleNotFou...
[  150s] FAILED distributed/tests/test_client.py::test_de_serialization_none - ModuleN...
[  150s] FAILED distributed/tests/test_pubsub.py::test_speed - ModuleNotFoundError: No...
[  150s] ERROR distributed/tests/test_nanny.py
[  150s] ERROR distributed/tests/test_utils.py
```

I am using this ad-hoc patch right now. It might need some polishing to meet your code standards:

[importorskip-numpy-tests.patch.txt](https://github.com/dask/distributed/files/6018277/importorskip-numpy-tests.patch.txt)

--- a/distributed/protocol/tests/test_arrow.py
+++ b/distributed/protocol/tests/test_arrow.py
@@ -1,7 +1,7 @@
-import pandas as pd
 import pytest
 
 pa = pytest.importorskip("pyarrow")
+pd = pytest.importorskip("pandas")
 
 import distributed
 from distributed.utils_test import gen_cluster

--- a/distributed/protocol/tests/test_collection.py
+++ b/distributed/protocol/tests/test_collection.py
@@ -1,7 +1,8 @@
 import pytest
 from distributed.protocol import serialize, deserialize
-import pandas as pd
-import numpy as np
+
+np = pytest.importorskip("numpy")
+pd = pytest.importorskip("pandas")
 
 
 @pytest.mark.parametrize("collection", [tuple, dict, list])

--- a/distributed/protocol/tests/test_collection_cuda.py
+++ b/distributed/protocol/tests/test_collection_cuda.py
@@ -2,7 +2,6 @@ import pytest
 
 from distributed.protocol import serialize, deserialize
 from dask.dataframe.utils import assert_eq
-import pandas as pd
 
 
 @pytest.mark.parametrize("collection", [tuple, dict])
@@ -37,14 +36,14 @@ def test_serialize_cupy(collection, y, y_serializer):
 @pytest.mark.parametrize("collection", [tuple, dict])
 @pytest.mark.parametrize(
     "df2,df2_serializer",
-    [(pd.DataFrame({"C": [3, 4, 5], "D": [2.5, 3.5, 4.5]}), "cuda"), (None, "pickle")],
+    [({"C": [3, 4, 5], "D": [2.5, 3.5, 4.5]}, "cuda"), (None, "pickle")],
 )
 def test_serialize_pandas_pandas(collection, df2, df2_serializer):
     cudf = pytest.importorskip("cudf")
-
+    pd = pytest.importorskip("pandas")
     df1 = cudf.DataFrame({"A": [1, 2, None], "B": [1.0, 2.0, None]})
     if df2 is not None:
-        df2 = cudf.from_pandas(df2)
+        df2 = cudf.from_pandas(pd.DataFrame(df2))
     if issubclass(collection, dict):
         header, frames = serialize(
             {"df1": df1, "df2": df2}, serializers=("cuda", "dask", "pickle")

--- a/distributed/protocol/tests/test_keras.py
+++ b/distributed/protocol/tests/test_keras.py
@@ -1,13 +1,14 @@
-import numpy as np
-from numpy.testing import assert_allclose
 import pytest
 
 keras = pytest.importorskip("keras")
+np = pytest.importorskip("numpy")
 
 from distributed.protocol import serialize, deserialize, dumps, loads, to_serialize
 
 
 def test_serialize_deserialize_model():
+    from numpy.testing import assert_allclose
+
     model = keras.models.Sequential()
     model.add(keras.layers.Dense(5, input_dim=3))
     model.add(keras.layers.Dense(2))

--- a/distributed/protocol/tests/test_numpy.py
+++ b/distributed/protocol/tests/test_numpy.py
@@ -1,7 +1,8 @@
 from zlib import crc32
 
-import numpy as np
 import pytest
+
+np = pytest.importorskip("numpy")
 
 from distributed.protocol import (
     serialize,

--- a/distributed/protocol/tests/test_pandas.py
+++ b/distributed/protocol/tests/test_pandas.py
@@ -1,6 +1,7 @@
-import numpy as np
-import pandas as pd
 import pytest
+
+pd = pytest.importorskip("pandas")
+np = pytest.importorskip("numpy")
 
 from dask.dataframe.utils import assert_eq
 

--- a/distributed/protocol/tests/test_serialize.py
+++ b/distributed/protocol/tests/test_serialize.py
@@ -6,7 +6,10 @@ import msgpack
 import pytest
 from tlz import identity
 
-np = pytest.importorskip("numpy")
+try:
+    import numpy as np
+except ImportError:
+    np = None
 
 from dask.utils_test import inc
 
@@ -213,6 +216,7 @@ def test_empty_loads_deep():
     assert isinstance(e2[0][0][0], Empty)
 
 
+@pytest.mark.skipif(np is None, reason="Test needs numpy")
 @pytest.mark.parametrize("kwargs", [{}, {"serializers": ["pickle"]}])
 def test_serialize_bytes(kwargs):
     for x in [
@@ -229,6 +233,7 @@ def test_serialize_bytes(kwargs):
         assert str(x) == str(y)
 
 
+@pytest.mark.skipif(np is None, reason="Test needs numpy")
 def test_serialize_list_compress():
     pytest.importorskip("lz4")
     x = np.ones(1000000)
@@ -440,7 +445,11 @@ def test_compression_numpy_list():
         (tuple([MyObj(None)]), True),
         ({("x", i): MyObj(5) for i in range(100)}, True),
         (memoryview(b"hello"), True),
-        (memoryview(np.random.random((3, 4))), True),
+        pytest.param(
+            memoryview(np.random.random((3, 4)) 
+                       if np is not None else b"skip np.random"),
+            True,
+            marks=pytest.mark.skipif(np is None, reason="Test needs numpy")),
     ],
 )
 def test_check_dask_serializable(data, is_serializable):
@@ -463,7 +472,14 @@ def test_serialize_lists(serializers):
 
 
 @pytest.mark.parametrize(
-    "data_in", [memoryview(b"hello"), memoryview(np.random.random((3, 4)))]
+    "data_in",
+    [
+        memoryview(b"hello"),
+        pytest.param(
+            memoryview(np.random.random((3, 4))
+                       if np is not None else b"skip np.random"),
+            marks=pytest.mark.skipif(np is None, reason="Test needs numpy"))
+    ],
 )
 def test_deser_memoryview(data_in):
     header, frames = serialize(data_in)
@@ -473,6 +489,7 @@ def test_deser_memoryview(data_in):
     assert data_in == data_out
 
 
+@pytest.mark.skipif(np is None, reason="Test needs numpy")
 def test_ser_memoryview_object():
     data_in = memoryview(np.array(["hello"], dtype=object))
     with pytest.raises(TypeError):

--- a/distributed/protocol/tests/test_serialize.py
+++ b/distributed/protocol/tests/test_serialize.py
@@ -3,9 +3,10 @@ import copy
 import pickle
 
 import msgpack
-import numpy as np
 import pytest
 from tlz import identity
+
+np = pytest.importorskip("numpy")
 
 from dask.utils_test import inc
 

--- a/distributed/protocol/tests/test_serialize.py
+++ b/distributed/protocol/tests/test_serialize.py
@@ -446,10 +446,12 @@ def test_compression_numpy_list():
         ({("x", i): MyObj(5) for i in range(100)}, True),
         (memoryview(b"hello"), True),
         pytest.param(
-            memoryview(np.random.random((3, 4)) 
-                       if np is not None else b"skip np.random"),
+            memoryview(
+                np.random.random((3, 4)) if np is not None else b"skip np.random"
+            ),
             True,
-            marks=pytest.mark.skipif(np is None, reason="Test needs numpy")),
+            marks=pytest.mark.skipif(np is None, reason="Test needs numpy")
+        ),
     ],
 )
 def test_check_dask_serializable(data, is_serializable):
@@ -476,9 +478,11 @@ def test_serialize_lists(serializers):
     [
         memoryview(b"hello"),
         pytest.param(
-            memoryview(np.random.random((3, 4))
-                       if np is not None else b"skip np.random"),
-            marks=pytest.mark.skipif(np is None, reason="Test needs numpy"))
+            memoryview(
+                np.random.random((3, 4)) if np is not None else b"skip np.random"
+            ),
+            marks=pytest.mark.skipif(np is None, reason="Test needs numpy")
+        ),
     ],
 )
 def test_deser_memoryview(data_in):

--- a/distributed/protocol/tests/test_serialize.py
+++ b/distributed/protocol/tests/test_serialize.py
@@ -450,7 +450,7 @@ def test_compression_numpy_list():
                 np.random.random((3, 4)) if np is not None else b"skip np.random"
             ),
             True,
-            marks=pytest.mark.skipif(np is None, reason="Test needs numpy")
+            marks=pytest.mark.skipif(np is None, reason="Test needs numpy"),
         ),
     ],
 )
@@ -481,7 +481,7 @@ def test_serialize_lists(serializers):
             memoryview(
                 np.random.random((3, 4)) if np is not None else b"skip np.random"
             ),
-            marks=pytest.mark.skipif(np is None, reason="Test needs numpy")
+            marks=pytest.mark.skipif(np is None, reason="Test needs numpy"),
         ),
     ],
 )

--- a/distributed/protocol/tests/test_sparse.py
+++ b/distributed/protocol/tests/test_sparse.py
@@ -1,13 +1,14 @@
-import numpy as np
-from numpy.testing import assert_allclose
 import pytest
 
+np = pytest.importorskip("numpy")
 sparse = pytest.importorskip("sparse")
 
 from distributed.protocol import deserialize, serialize
 
 
 def test_serialize_deserialize_sparse():
+    from numpy.testing import assert_allclose
+
     x = np.random.random((2, 3, 4, 5))
     x[x < 0.8] = 0
 

--- a/distributed/tests/test_actor.py
+++ b/distributed/tests/test_actor.py
@@ -425,8 +425,8 @@ async def test_load_balance_map(c, s, *workers):
 
 @gen_cluster(client=True, nthreads=[("127.0.0.1", 1)] * 4, Worker=Nanny)
 async def bench_param_server(c, s, *workers):
+    np = pytest.importorskip("numpy")
     import dask.array as da
-    import numpy as np
 
     x = da.random.random((500000, 1000), chunks=(1000, 1000))
     x = x.persist()

--- a/distributed/tests/test_client.py
+++ b/distributed/tests/test_client.py
@@ -5535,7 +5535,7 @@ async def test_client_active_bad_port():
 def test_turn_off_pickle(direct):
     @gen_cluster()
     async def test(s, a, b):
-        import numpy as np
+        np = pytest.importorskip("numpy")
 
         async with Client(
             s.address, asynchronous=True, serializers=["dask", "msgpack"]
@@ -5573,7 +5573,7 @@ def test_turn_off_pickle(direct):
 
 @gen_cluster()
 async def test_de_serialization(s, a, b):
-    import numpy as np
+    np = pytest.importorskip("numpy")
 
     c = await Client(
         s.address,
@@ -5594,7 +5594,7 @@ async def test_de_serialization(s, a, b):
 
 @gen_cluster()
 async def test_de_serialization_none(s, a, b):
-    import numpy as np
+    np = pytest.importorskip("numpy")
 
     c = await Client(s.address, asynchronous=True, deserializers=["msgpack"])
     try:
@@ -6432,8 +6432,8 @@ async def test_annotations_retries(c, s, a, b):
 @gen_cluster(client=True)
 async def test_annotations_blockwise_unpack(c, s, a, b):
     da = pytest.importorskip("dask.array")
+    np = pytest.importorskip("numpy")
     from dask.array.utils import assert_eq
-    import numpy as np
 
     # A flaky doubling function -- need extra args because it is called before
     # application to establish dtype/meta.

--- a/distributed/tests/test_collections.py
+++ b/distributed/tests/test_collections.py
@@ -2,8 +2,8 @@ from distutils.version import LooseVersion
 
 import pytest
 
-pytest.importorskip("numpy")
-pytest.importorskip("pandas")
+np = pytest.importorskip("numpy")
+pd = pytest.importorskip("pandas")
 
 import dask
 import dask.dataframe as dd
@@ -11,8 +11,6 @@ import dask.bag as db
 from distributed.client import wait
 from distributed.utils_test import gen_cluster
 from distributed.utils_test import client, cluster_fixture, loop  # noqa F401
-import numpy as np
-import pandas as pd
 
 PANDAS_VERSION = LooseVersion(pd.__version__)
 PANDAS_GT_100 = PANDAS_VERSION >= LooseVersion("1.0.0")

--- a/distributed/tests/test_nanny.py
+++ b/distributed/tests/test_nanny.py
@@ -7,8 +7,6 @@ import random
 import sys
 import multiprocessing as mp
 
-import numpy as np
-
 import pytest
 from tlz import valmap, first
 from tornado.ioloop import IOLoop
@@ -202,6 +200,7 @@ async def test_random_seed(c, s, a, b):
         assert x != y
 
     await check_func(lambda a, b: random.randint(a, b))
+    np = pytest.importorskip("numpy")
     await check_func(lambda a, b: np.random.randint(a, b))
 
 

--- a/distributed/tests/test_pubsub.py
+++ b/distributed/tests/test_pubsub.py
@@ -37,7 +37,7 @@ async def test_speed(c, s, a, b):
             #     print(a, b, i)
         return n
 
-    import numpy as np
+    np = pytest.importorskip("numpy")
 
     x = np.random.random(1000)
 

--- a/distributed/tests/test_utils.py
+++ b/distributed/tests/test_utils.py
@@ -10,7 +10,6 @@ import sys
 from time import sleep
 import traceback
 
-import numpy as np
 import pytest
 from tornado.ioloop import IOLoop
 
@@ -271,6 +270,7 @@ def test_ensure_bytes():
 
 
 def test_ensure_bytes_ndarray():
+    np = pytest.importorskip("numpy")
     result = ensure_bytes(np.arange(12))
     assert isinstance(result, bytes)
 
@@ -283,6 +283,8 @@ def test_ensure_bytes_pyarrow_buffer():
 
 
 def test_nbytes():
+    np = pytest.importorskip("numpy")
+
     def check(obj, expected):
         assert nbytes(obj) == expected
         assert nbytes(memoryview(obj)) == expected

